### ISSUE TITLE
[MULAN] Will the images per batch influence the detection accuracy?

### ISSUE THREAD

**Issue by Dalerting** [rsummers11/CADLab#29]
Recently I try to reproduce the result of your paper. I set the value of images per batch to 2 and other value remains, and then the mean of detection accuracy is just about 81%. 

The three branch are all set to be true and additional feature is added too. when the value of image per batch is 3 ,the accuracy will slight increase. I didn't continue to increase the value due to the limitation of my GPU. However, the tag accuracy and the segmentation accuracy are close to the result of your paper whatever the value is. 

Should I change the learning rate or something if the images per batch is different from yours?

Here are the parts of my experiment result,and the images per batch is 2.Thank you for your help!

1: 0.5968
2: 0.6065
3: 0.5630
4: 0.6423
5: 0.7836
6: 0.7941
7: 0.8005
8: 0.8003

Detection accuracy:
Sensitivity @ [0.5, 1, 2, 4, 8, 16] average FPs per image:
0.6932, 0.7852, 0.8571, 0.8998, 0.9281, 0.9471
mean of [0.5, 1, 2, 4]: 0.8088

Tagging accuracy:
hand-labeled tags:
m_AUC	pc_F1	pc_Pr	pc_Re	wm_AUC	ov_F1	ov_Pr	ov_Re	
0.9490	0.4588	0.4437	0.6207	0.9573	0.6270	0.5323	0.7625	

Segmentation accuracy:
avg min distance (mm) from groundtruth recist points to predicted contours in GT boxes:
error of lesion diameter (mm) estimated from predicted contours in GT boxes:
1.4376+-1.6287, 1.9919+-2.2651
total test time: 666.962127

**Comment by viggin** [comment]
I haven't tried to reduce batch size to 2. Theoretically, lr should be reduced proportionally to batch size. So you can try to change lr.
You can also try to modify the code to cumulate gradients of multiple batches.

**Comment by Dalerting** [comment]
Thank you for your reply. I have tried different learning rate and the way of accumulating gradients, but it didn't work. The detection accuracy was almost same as before or worse.

 I tried to run your project on another PC. The training progress was fine.However there comes an assertion error in the evaluation.
 assert lb_te_all.shape == pred_te_all.shape (in the tagging_eval.py)
The lb_te_all.shape is 4791 and pred_te_all.shape is about 5800 or so. It confused me and I haven't find out the solution.

 And then I try to ablate the tagging branch.The program runs well and it does increase and achieve about 84.8% when the tagging branch was not added. That is close the result of your paper. I think there may be some relations between the detection branch and tagging branch as you said in your paper.

Please help me to solve the problem. Thank you !

**Comment by viggin** [comment]
I couldn't find a reason why pred_te_all.shape is about 5800 or so.

**Comment by Dalerting** [comment]
Is so strange that the program runs well now when I want to print the shape. The param pred_te_all is refered to the pred_all in the DL_eval.py

    pred_all = [predictions[fn]['gt_result'].get_field('tag_predictions') for fn in fns]
    print("before:",len(pred_all))

    pred_all = torch.cat(pred_all).cpu().numpy()
    print("after:", len(pred_all))

The outputs are following:
before:4791
after:4887

I have no idea why such thing happened and the assertion error is gone. Anyway, it runs well now. Thank you !

**Comment by AxGa95** [comment]
@Dalerting Hi, im running into the same problem with assert lb_te_all.shape == pred_te_all.shape can you tell me how did you  ablate the tagging branch? I`m new to this, thank you!
